front/lib/api/assistant/generation.ts

```diff
@@ -235,12 +238,15 @@ export async function renderConversationForModel({
     },
     "[ASSISTANT_TRACE] Genration message token counts for model conversation rendering"
   );
+  while (selected.length > 0 && selected[0].message.role === "agent") {
+    selected.shift();
+  }
 
   return new Ok({
     modelConversation: {
-      messages: selected,
+      messages: selected.map((s) => s.message),
     },
-    tokensUsed,
+    tokensUsed: selected.map((s) => s.tokenUseds).reduce((a, b) => a + b, 0),
```

Comment:
super nit but you could maintain tokensUsed as you shift no ?

Reply:
Indeed, if you prefer it that way, I can change it :)